modules/core/src/smithy4s/http/internals/SchemaVisitorMetadataReader.scala

```diff
@@ -0,0 +1,304 @@
+  StructureMetaDecode
+}
+import smithy4s.internals.InputOutput
+import smithy4s.schema._
+
+import java.{util => ju}
+import scala.collection.mutable.{ListBuffer, Map => MMap}
+
+private[http] object SchemaVisitorMetadataReader
+    extends SchemaVisitorMetadataReader
+
+private[http] class SchemaVisitorMetadataReader()
+    extends SchemaVisitor[MetaDecode]
+    with ScalaCompat { self =>
+
+  override def primitive[P](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: Primitive[P]
+  ): MetaDecode[P] = {
+    tag match {
+      case Primitive.PShort  => MetaDecode.from("Short")(_.toShortOption)
+      case Primitive.PInt    => MetaDecode.from("Int")(_.toIntOption)
+      case Primitive.PFloat  => MetaDecode.from("Float")(_.toFloatOption)
+      case Primitive.PLong   => MetaDecode.from("Long")(_.toLongOption)
+      case Primitive.PDouble => MetaDecode.from("Double")(_.toDoubleOption)
+      case Primitive.PBigInt => MetaDecode.fromUnsafe("BigInt")(BigInt(_))
+      case Primitive.PBigDecimal =>
+        MetaDecode.fromUnsafe("BigDecimal")(BigDecimal(_))
+      case Primitive.PBoolean => MetaDecode.from("Boolean")(_.toBooleanOption)
+      case Primitive.PString  => MetaDecode.fromUnsafe("String")(identity)
+      case Primitive.PUUID =>
+        MetaDecode.fromUnsafe[ju.UUID]("UUID")(ju.UUID.fromString)
+      case Primitive.PByte => MetaDecode.EmptyMetaDecode
+      case Primitive.PBlob =>
+        MetaDecode.fromUnsafe("Bytes")(string =>
+          ByteArray(ju.Base64.getDecoder().decode(string))
+        )
+      case Primitive.PDocument => MetaDecode.EmptyMetaDecode
+      case Primitive.PTimestamp =>
+        (
+          hints.get(HttpBinding).map(_.tpe),
+          hints.get(smithy.api.TimestampFormat)
+        ) match {
+          case (_, Some(format)) =>
+            MetaDecode.from(Timestamp.showFormat(format))(str =>
+              Timestamp.parse(str, format)
+            )
+          case (Some(HttpBinding.Type.QueryType), None) |
+              (Some(HttpBinding.Type.PathType), None) =>
+            val formatString =
+              Timestamp.showFormat(smithy.api.TimestampFormat.DATE_TIME)
+            // See https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html?highlight=httpquery#httpquery-trait
+            MetaDecode.from(formatString)(str =>
+              Timestamp.parse(str, smithy.api.TimestampFormat.DATE_TIME)
+            )
+          case (Some(HttpBinding.Type.HeaderType), None) =>
+            val formatString =
+              Timestamp.showFormat(smithy.api.TimestampFormat.HTTP_DATE)
+            // See https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html?highlight=httpquery#httpheader-trait
+            MetaDecode.from(formatString)(str =>
+              Timestamp.parse(str, smithy.api.TimestampFormat.HTTP_DATE)
+            )
+          case (None, None) =>
+            MetaDecode.EmptyMetaDecode
+        }
+      case Primitive.PUnit =>
+        MetaDecode.StructureMetaDecode(
```

Comment:
for my knowledge, is this different than `EmptyMetaDecode` ?

Reply:
yes, afaict, this is used for cases where the input is Unit